### DISCUSSION TITLE
Multi-class selector in component layer not ordered correctly?

### DISCUSSION THREAD

**Discussion by Anders2303** [tailwindlabs/tailwindcss#12860] [Help]
Technical info:
* Tailwind version: 3.3.1

If I write a component layer styling that's based on two (or more) classes, the multi-class style is applied _after_ the tailwind utilities, so I can no longer override them. Is this intentional by tailwind, or a bug?

For an example, say my components are defined as follows:
```css
@layer components {
  .my-button {
    @apply rounded-md p-2;
  }

  .my-button.--dangerous {
    @apply bg-red-500;
  }

  /* ... other my-button colors */
}
```

If I then write my button as `<button class="my-button --dangerous bg-purple-200"> ... </button>` the button will _not_ get the purple background, since the `.my-button.--dangerous` styling is applied _after_ the tailwind utilities. From the docs, I understood that all styling written inside the components layer should be ordered before the utilities layer, so why is this happening?

![2024-01-31-105836_grim](https://github.com/tailwindlabs/tailwindcss/assets/23122793/85b4d8c5-3ecb-417f-86d7-732f422e807a)

**Comment by wongjn** [comment]
It seems you have two CSS files, `buttons.css` and `utilities.css`. if these are indeed two separate files when they reach the browser, consider consolidating them into once CSS file so that the `@layer` rules can be injected in the correct order.

**Reply by adamwathan** [reply]
In addition to this, the selector `.my-button.--dangerous` has a higher specificity than `.bg-purple-500` because it contains two classes.

You could solve this by putting things into native cascade layers, which Tailwind currently doesn't do automatically as our `@layer` feature existed before native `@layer` existed in CSS (something we're changing in V4 though).

**Reply by Anders2303** [reply]
@wongjn The `utilities.css` is the utility-layer from tailwindcss! So I cant consolidate them. In this example, the only file I have is `buttons.css`.

@adamwathan Thanks for the suggestion, I'll take a look at native layers. A followup question regarding the selector: is the two-class selector being placed after the tailwind-classes just something that happens in native CSS? Will selectors with higher specificity ALWAYS be applied after tailwind's single-class selectors?
